### DISCUSSION TITLE
How to create multiple password rules based on user input?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1024] [Q&A] [answered]
### Question

Hi All,

I am trying to create a password input with multiple validation rules, and show some labels below. As and when each rule is satisfied, the label gets updated (maybe color changes, or there is a strike-through in the text). 
Something like this image below:
![image](https://github.com/zauberzeug/nicegui/assets/65127974/7d2c17e9-2939-4294-940b-e1ac98596ba6)

I assume, the latter part I can create with a ui.card() and some icons and labels etc. But first, I am trying to figure out how to check for each rule and then connect that pass/fail condition with the specific label/icon. 

So for ex: I created the basic password  button, rules, and then displayed the rules in labels below like this:
```python
# Show password input
pwd = ui.input("Insert Password", password=True)

# Validation rules
rule1 = len(pwd.value) >= 8                                     # 8 characters
rule2 = re.search(r"[A-Z]", pwd.value)                          # at least one uppercase
rule3 = re.search(r"[a-z]", pwd.value)                          # at least one lowercase
rule4 = re.search(r"\d", pwd.value)                             # at least one digit
rule5 = re.search(r"[!@#$%^&*(),.?':{}|<>]", pwd.value)         # at least one special character

ui.label(f'Rule1 is {rule1}')
ui.label(f'Rule2 is {rule2}')
ui.label(f'Rule3 is {rule3}')
ui.label(f'Rule4 is {rule4}')
ui.label(f'Rule5 is {rule5}')
```

But since the rules are not connected to the pwd input, hence it's not checking anything. I assume I need to use the `on_change` event to check for these rules, but how do I do this for all 5 of them together?

Just to explain a bit more, what I am really trying to achieve is something like, let's say in the beginning the label for first rule (8 character) is like this:
```python
with ui.row().classes('items-center'):
    ui.icon('radio_button_unchecked').props('color=red')
    ui.label("At least 8 characters").classes('text-xs text-red')
```

If the rule is satisfied, I want it to become like this:
```python
with ui.row().classes('items-center'):
    ui.icon('done').props('color=green')
    ui.label("At least 8 characters").classes('text-xs text-green strike-through')  # strike out means it's satisfied
```

I tried understanding from the other existing examples, but can't figure out how to use the ```on_change``` and ```bind_visibility``` properties properly to do this. Please advise me on the proper way to do this. If the thinking is wrong all the way, then please let me know how else should I approach this. Really appreciate your help on this.

Thanks,
Anindya

**Comment by falkoschindler** [comment, marked as answer]
That's an interesting problem, @Anindya088!
Although it might be possible to bind the labels' visibility, I tend to simply use a piece of "refreshable" UI. It can re-create the labels whenever the input changes:

```py
pwd = ui.input('Password', password=True, on_change=lambda: show_info.refresh())

rules = {
    '# of characters': lambda s: len(s) >= 8,
    'Lowercase letter': lambda s: re.search(r'[a-z]', s),
    'Capital letter': lambda s: re.search(r'[A-Z]', s),
    'Number': lambda s: re.search(r'\d', s),
    'Special character': lambda s: re.search(r"[!@#$%^&*(),.?':{}|<>]", s),
}

@ui.refreshable
def show_info():
    for rule, check in rules.items():
        with ui.row().classes('items-center'):
            if check(pwd.value or ''):
                ui.icon('done', color='green')
                ui.label(rule).classes('text-xs text-green strike-through')
            else:
                ui.icon('radio_button_unchecked', color='red')
                ui.label(rule).classes('text-xs text-red')

show_info()
```

**Reply by Anindya088** [reply]
@falkoschindler , this is absolutely amazing, you make it look so simple and easy :smiley:

I did not know about the "refreshable" and then connecting it back to ```on_change```, time for me to go back and read more on the docs.

Thanks,
Anindya
